Glenn3

glennfrutiz opened the pull request:
added search bar to search page.
added route for when a user makes a search
fixed query used in that route to look for substrings in title, genre, or author
updated links/buttons to redirect to /faq instead of /pointsFAQ

glennfrutiz committed:
Create pointsFAQ.handlebars

glennfrutiz committed:
Feature 1: Request a book

Integrated my features into Andrew's recent pull request.
*****in public\client\search.js*******
1. updated requestBook function to prevent requests when the user doesn't have enough points

2. added call to googleapi route created by Melissa to populate book details into a modal
*****in routes\search.js*******
Update select all books available using current users location and whether a seller will ship worldwide (all available books from seller same country + books from seller other country that will allow worldwide)
*****added views\pointsFAQ.handlebars*******
rudimentary page containing basic rules for points
*****in views\search.handlebars*******
1. added link to pointsFAQ
2. added functionality to "View Details" button provided by Andrew through the inclusion of a modal for book details
3, prevented user from making request  if they don't have enough points on "Request a copy" and displayed message

glennfrutiz committed:
Changes based on comments

glennfrutiz committed:
Changes based on comments

glennfrutiz committed:
cleaned up modal

glennfrutiz committed:
Update search.js

cleaned up and changed request functions to handle errors caused by books with apostrophes in titles

glennfrutiz committed:
Update search.js

Added rough draft for search results router

glennfrutiz committed:
Update search.handlebars

added quotes to onclick functions to avoid errors from books with apostrophes or quotes in title

glennfrutiz committed:
Update main.handlebars

add action and method to form containing search bar

glennfrutiz committed:
Merge branch 'master' into glenn3

glennfrutiz committed:
debugging some things

updated common function to get pending points to reflect recent changes to database. removed points FAQ and updated search page to lead to faq instead.

glennfrutiz committed:
more debugging

changed router to update new name for faq

glennfrutiz committed:
Merge branch 'master' into glenn3

glennfrutiz committed:
search bar bug fixes

moved search bar to search page
fixed bugs in search query
changed setting innerHTML of pageModal.

glennfrutiz commented:
please do

glennfrutiz commented:
working on filtering search results. will send pull request when complete

glennfrutiz committed:
filtering

filtering implemented on search by by author, title, or genre, rating, publication date

glennfrutiz committed:
Update search.js

filtering wasn't working after making a search. updated query to make it work.

glennfrutiz commented:
Reopened request to include filter functionality on search page. It toggles cards off when they do not contain a string the user types into a searchbar. The user can decide where they want to look for the string. For example either in the title, pubDate, rating, etc.
I was lucky to find this codepen last night that used filtering with cards: https://codepen.io/r3hab/pen/GBdXEq

While making this i had an idea. Why not scrap searching information in the database using a toolbar. Instead let's display all available books and use the filtering function to perform the search on the cards. I recorded snippet of it on https://www.youtube.com/watch?v=EDUpML6MoDc

P.S. need to improve styling next week. I also noticed it takes a while to load upon login. I think this might be because i made the SQL queries larger to get information needed to filter.

mkoeplin commented on common.js:
```diff
@@ -40,11 +40,11 @@ function getPoints(userId, callback) {
     WHERE requestorId = ? AND statusId <> 8 
     UNION 
-    SELECT SUM(t.pointCost) AS points 
+    SELECT SUM(t.buyerPoints) AS points 
```
I think this should be sellerPoints. It shouldn't make a huge difference in the results but since this select statement is pulling books where they are the seller we should use sellerPoints.

mkoeplin commented on public/client/search.js:
```diff
@@ -7,17 +7,18 @@ function requestBook(bookId, pointCost, title, userPoints){
         var confirmButton = document.getElementById('confirmButton');
-        confirmButton.setAttribute("onclick", `request(${bookId}, ${pointCost}, '${title}')`);
+        //title was previously surrounded in single quotes
+        confirmButton.setAttribute("onclick", `request(${bookId}, ${pointCost}, "${title}")`);
```
This button still shouldn't be done in the js. I noted this in the last pull request that the button should just be in the modal with an onclick as I don't think it's dynamic and that's how other buttons are created in our code base.

mkoeplin commented on routes/search.js:
```diff
@@ -23,16 +26,20 @@ router.get('/', common.isAuthenticated, (req, res, next) => {
     })
 })
 
+router.get('/faq', common.isAuthenticated, (req,res,next) => {
```
I removed this from search and put it in index in my last PR. I don't think this should be in the search file as it ends with a weird route of "/search/faq" where as in the index file it will display as "/faq". Plus this page is not specific to search, it's entire application relevant.

mkoeplin commented on routes/search.js:
```diff
@@ -23,16 +26,20 @@ router.get('/', common.isAuthenticated, (req, res, next) => {
 // ******** need to update this based on worldwide shippers
 function allBooks(id, callback){
-    var selectAllAvailableBooks = `SELECT UserBooks.id AS id, title AS title, author AS author, imgUrl AS imgUrl, isbn13 AS isbn, Users.country
+    var selectAllAvailableBooks = `SELECT UserBooks.id AS id, title AS title, author AS author, imgUrl AS imgUrl, isbn13 AS isbn, genre AS genre, rating AS rating, pubDate AS pubDate, Users.country
```
We shouldn't need the AS parts of this on every column you select. That is only needed if you want to change the column return name OR you are unioning two selects with different column names in which case you would only apply it to the column names that are different.

mkoeplin commented on routes/search.js:
```diff
@@ -23,16 +26,20 @@ router.get('/', common.isAuthenticated, (req, res, next) => {
                                  WHERE UserBooks.userId != ? AND UserBooks.available = 1 AND Users.worldwide = 1
                                  UNION
-                                 SELECT UserBooks.id AS id, title AS title, author AS author, imgUrl AS imgUrl, isbn13 AS isbn, Users.country
+                                 SELECT UserBooks.id AS id, title AS title, author AS author, imgUrl AS imgUrl, isbn13 AS isbn, genre AS genre, rating AS rating, pubDate AS pubDate, Users.country
```
Same as above, I don't think we need the AS parts of this query.

mkoeplin commented on routes/search.js:
```diff
@@ -142,4 +152,87 @@ router.post('/', common.isAuthenticated, (req, res, next) => {
+    
+    
+    var selectBooks = `SELECT UserBooks.id AS id, title AS title, author AS author, imgUrl AS imgUrl, isbn13 AS isbn, genre AS genre, rating AS rating, pubDate AS pubDate, Users.country
+                                 FROM UserBooks
```
Same as above, we likely don't need the AS parts of the select

mkoeplin commented on views/search.handlebars:
```diff
@@ -41,7 +71,7 @@
             </div>
             <div class="modal-body" id="modalBody">
-
+                Are you sure you want this book?: <br><b><p id="requestTitle"></p></b>
```
Bold tags are not bootstrap. The p tag should have like class="font-weight-bold" on it

mkoeplin commented on views/search.handlebars:
```diff
@@ -134,5 +164,73 @@
+                </button>
+            </div>
+            <div class="modal-body" id="modalBody">
+                <b><h8>Title</h8></b><br>
```
These should probably be adjusted to use bootstrap.

mkoeplin reviewed (changes requested):
As discussed, since no one was aware you decided to do search and the implementation was not discussed we need to come together as a group to determine whether this approach fits the heuristics and design. I left some other comments as well on cleanup I think should be done. Since it's the last day of sprint, I don't believe we should move this in and would suggest this PR be abandoned and pushed once changes are made and the design is agreed to.

mkoeplin commented on views/search.handlebars:
```diff
@@ -7,22 +7,52 @@
-                    <p class="card-text mb-0">{{title}}</p>
-                    <p class="card-text">{{author}}</p>
+                    <p class="card-text mb-0 card-title">{{title}}</p>
+                    <p class="card-text mb-0 card-author">{{author}}</p>
```
You don't need to have styling on the hidden fields. I'd remove that and just have these be hidden divs not even paragraphs as paragraphs have their own styling that could mess with things.

glennfrutiz committed:
client side search

applied suggestions and went with client side search.

removed search bar that depended on a router.

glennfrutiz committed:
Some fixes based on comments

glennfrutiz committed:
Merge branch 'master' into glenn3

glennfrutiz committed:
more changes based on comments

Removed modal for confirmation of book request. replaced with confirm element that triggers request function. Added flash message to be displayed after book has been requested.

Applied initial bootstrap styling to modals.

mkoeplin reviewed (approved):
Nice job with the changes, can't wait to try out the new client-side search!
